### PR TITLE
[train] Hard-deprecate `MosaicTrainer` and remove `SklearnTrainer`

### LINKED ISSUE #32732: [AIR][Train] Multilabel classification with `SklearnTrainer`
### Description

Here in ray 2.x, we have the sklearnTrainer support. 
Q1: does it support multilabel classification approach like OneVsRestClassifier?
SklearnTrainer class have one argument in the constructor like label_column. Does it take list of labels there? I tried but it is not working.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

`MosaicTrainer` is not needed, since it can be folded under `TorchTrainer`, similar to [`LightningTrainer` and `TransformersTrainer`](https://github.com/ray-project/enhancements/blob/main/reps/2023-07-20-torch-trainer-apis.md)

`SklearnTrainer` is also not needed, since it does not provide any extra utility over just calling `Tuner(sklearn_train_fn)`. The Ray Train version added an extra layer of abstraction that is not needed over calling the underlying sklearn methods. There was no multi-worker training happening.

### TODO

- [ ] Finish writing the migration issue.

### Notes

* [Mosaic Composer does not actually work with Ray Train as of 2.9.1](https://github.com/ray-project/ray/issues/40035) due to a signal handler it tries to register that must run on the main thread. Ray Train currently creates a worker thread to execute the training logic. This is because Ray Train implements pause-able training execution for each training worker by communicating between a main thread and a worker thread through `ray.train.report`.

## Related issue number

Closes https://github.com/ray-project/ray/issues/32732

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/mosaic/mosaic_trainer.py [hunk 1]
```diff
@@ -1,255 +1,30 @@
...
-from typing import TYPE_CHECKING, Any, Callable, Dict, Optional, Type
+from ray.util.annotations import Deprecated
 
-from composer.loggers.logger_destination import LoggerDestination
-from composer.trainer import Trainer
+_DEPRECATION_MESSAGE = (
+    "`ray.train.mosaic.MosaicTrainer` is deprecated. "
+    "Use `ray.train.torch.TorchTrainer` instead. "
+    "See this issue for a migration example: "
+    "https://github.com/ray-project/ray/issues/42257"
```
Comment (matthewdeng) [inline python/ray/train/mosaic/mosaic_trainer.py:7]: We won't be able to have a migration example because of the threading issue though, right?

File: python/ray/train/sklearn/sklearn_predictor.py [hunk 2]
```diff
@@ -9,14 +9,29 @@
...
+        for param in estimator.get_params(deep=True)
+        if any(
+            param.endswith(cpu_param_name) for cpu_param_name in SKLEARN_CPU_PARAM_NAMES
+        )
+    }
+    estimator.set_params(**cpu_params)
+
+
 @PublicAPI(stability="alpha")
 class SklearnPredictor(Predictor):
```
Comment (woshiyyya) [inline python/ray/train/sklearn/sklearn_predictor.py:36]: Do we still need SklearnPredictor?

File: python/ray/train/mosaic/mosaic_trainer.py (hunk 1 above)
Comment (justinvyu) [reply python/ray/train/mosaic/mosaic_trainer.py:7]: Yeah, I was planning to just keep it as "not possible for now", then update the issue later. Or should I just remove the migration link for Mosaic trainer?

File: python/ray/train/sklearn/sklearn_predictor.py (hunk 2 above)
Comment (justinvyu) [reply python/ray/train/sklearn/sklearn_predictor.py:36]: We still have all the predictors around but not shown in docs. Maybe we can remove them all at some point at once?

**Comment by woshiyyya** [comment]
Does the main thread refer to the driver and all Ray actors methods are child threads? Seems that the signal handling issue is unresolvable in current Ray Train design.

**Comment by justinvyu** [comment]
@woshiyyya It refers to the main thread that is processing the result queue from the training thread:
* https://github.com/ray-project/ray/blob/master/python/ray/train/_internal/session.py#L203
* https://github.com/ray-project/ray/blob/master/python/ray/train/_internal/session.py#L274

> Seems that the signal handling issue is unresolvable in current Ray Train design.

Yep this is unresolvable unless we get rid of this threading logic (which is an implementation detail for implementing "yielding" behavior, that can be done with Ray Generators in the future).

**Comment by woshiyyya** [comment]
Got it. Signal handling is an important feature that we need to support it in the future.
- Lightning's `on_exception` [callback](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.callbacks.OnExceptionCheckpoint.html)
- [An oss issue](https://github.com/ray-project/ray/issues/39055)

File: python/ray/train/mosaic/mosaic_trainer.py (hunk 1 above)
Comment (matthewdeng) [reply python/ray/train/mosaic/mosaic_trainer.py:7]: Oh yeah we can have a GH Issue that says it's not supported right now (the current one points to sklearn)

**Review by matthewdeng — APPROVED** [review]
